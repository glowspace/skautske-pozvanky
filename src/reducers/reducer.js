import { map, path, assocPath, last, inc } from 'ramda';

export const changeSelected = (state, id) => {
  return assocPath(
    ['things'],
    map(thing => {
      if (path(['id'], thing) === id) {
        return assocPath(['selected'], !path(['selected'], thing), thing);
      }
      return thing;
    }, path(['things'], state)),
    state
  );
};

export const addThing = (state, name) => {
  return assocPath(
    ['things'],
    [
      ...path(['things'], state),
      {
        id: inc(path(['id'], last(path(['things'], state)))),
        selected: true,
        name
      }
    ],
    state
  );
};

export const changeText = (state, value) => {
  return assocPath(
    ['texts'],
    map(textBox => {
      if (path(['id'], textBox) === value.id) {
        return assocPath(['text'], path(['value'], value), textBox);
      }
      return textBox;
    }, path(['texts'], state)),
    state
  );
};

export const switchToBw = (state) => {
  return assocPath(['bw'], !path(['bw'], state), state)
}

const reducer = (state, action) => {
  switch (action.type) {
    case 'changeSelected':
      return changeSelected(state, action.id);
    case 'addThing':
      return addThing(state, action.name);
    case 'changeText':
      return changeText(state, action.value);
    case 'switchToBw':
      return switchToBw(state)
    default:
      return state;
  }
};

export default reducer;
