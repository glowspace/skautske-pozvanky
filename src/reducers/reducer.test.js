import { changeSelected, addThing, changeText, switchToBw } from './reducer';
import initialState from '../../assets/typeOne';

test('changeSelected', () => {
  expect(changeSelected({ ...initialState }, 1)).toMatchSnapshot();
});

test('add new thing to list of things', () => {
  expect(addThing({ ...initialState }, 'test')).toMatchSnapshot();
});

test('change  text by id', () => {
  expect(
    changeText(
      { ...initialState },
      {
        id: 1,
        value: 'changed'
      }
    )
  ).toMatchSnapshot();
});

test('switch to bw', () => {
  expect(switchToBw({ bw: true })).toEqual({ bw: false })
})