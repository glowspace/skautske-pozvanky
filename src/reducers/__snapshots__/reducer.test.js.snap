// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`add new thing to list of things 1`] = `
Object {
  "bw": false,
  "texts": Array [
    Object {
      "color": "rgb(154, 21, 143)",
      "id": 2,
      "markdown": true,
      "name": "Nadpis",
      "text": "Pozvánka na vánoční výpravu",
      "type": "header",
    },
    Object {
      "color": "rgb(154, 21, 143)",
      "id": 1,
      "markdown": true,
      "name": "Obsah",
      "text": "Vánoce už jsou za dveřmi! Proto se opět vydáme na vánoční výpravu do Nové Vsi pod Horou Svatého Šebestiána, která se bude konat **16. 12.–18. 12.**. Sejdeme se **na nádraží** v pátek 16. 12. **v 18:15** (vlak odjíždí v 18:40). Děti je třeba dopravit až na nádraží. Vrátíme se v neděli kolem **13:48** tamtéž.",
      "type": "normal",
    },
    Object {
      "color": "rgb(154, 21, 143)",
      "id": 3,
      "markdown": false,
      "name": "Patička",
      "text": "Janča, Domča a Lenča, vedoucí oddílu Sedmikrásky",
      "type": "footer",
    },
  ],
  "things": Array [
    Object {
      "id": 1,
      "name": "přezuvky",
      "selected": true,
    },
    Object {
      "id": 2,
      "name": "spacák",
      "selected": false,
    },
    Object {
      "id": 3,
      "name": "ešus",
      "selected": false,
    },
    Object {
      "id": 4,
      "name": "karimatku",
      "selected": false,
    },
    Object {
      "id": 5,
      "name": "věci na schůzku",
      "selected": false,
    },
    Object {
      "id": 6,
      "name": "zápisník a tužku",
      "selected": false,
    },
    Object {
      "id": 7,
      "name": "kroj",
      "selected": false,
    },
    Object {
      "id": 8,
      "name": "hygienu",
      "selected": false,
    },
    Object {
      "id": 9,
      "name": "baterku",
      "selected": false,
    },
    Object {
      "id": 10,
      "name": "večeři",
      "selected": false,
    },
    Object {
      "id": 11,
      "name": "šátek",
      "selected": false,
    },
    Object {
      "id": 12,
      "name": "test",
      "selected": true,
    },
  ],
}
`;

exports[`change  text by id 1`] = `
Object {
  "bw": false,
  "texts": Array [
    Object {
      "color": "rgb(154, 21, 143)",
      "id": 2,
      "markdown": true,
      "name": "Nadpis",
      "text": "Pozvánka na vánoční výpravu",
      "type": "header",
    },
    Object {
      "color": "rgb(154, 21, 143)",
      "id": 1,
      "markdown": true,
      "name": "Obsah",
      "text": "changed",
      "type": "normal",
    },
    Object {
      "color": "rgb(154, 21, 143)",
      "id": 3,
      "markdown": false,
      "name": "Patička",
      "text": "Janča, Domča a Lenča, vedoucí oddílu Sedmikrásky",
      "type": "footer",
    },
  ],
  "things": Array [
    Object {
      "id": 1,
      "name": "přezuvky",
      "selected": true,
    },
    Object {
      "id": 2,
      "name": "spacák",
      "selected": false,
    },
    Object {
      "id": 3,
      "name": "ešus",
      "selected": false,
    },
    Object {
      "id": 4,
      "name": "karimatku",
      "selected": false,
    },
    Object {
      "id": 5,
      "name": "věci na schůzku",
      "selected": false,
    },
    Object {
      "id": 6,
      "name": "zápisník a tužku",
      "selected": false,
    },
    Object {
      "id": 7,
      "name": "kroj",
      "selected": false,
    },
    Object {
      "id": 8,
      "name": "hygienu",
      "selected": false,
    },
    Object {
      "id": 9,
      "name": "baterku",
      "selected": false,
    },
    Object {
      "id": 10,
      "name": "večeři",
      "selected": false,
    },
    Object {
      "id": 11,
      "name": "šátek",
      "selected": false,
    },
  ],
}
`;

exports[`changeSelected 1`] = `
Object {
  "bw": false,
  "texts": Array [
    Object {
      "color": "rgb(154, 21, 143)",
      "id": 2,
      "markdown": true,
      "name": "Nadpis",
      "text": "Pozvánka na vánoční výpravu",
      "type": "header",
    },
    Object {
      "color": "rgb(154, 21, 143)",
      "id": 1,
      "markdown": true,
      "name": "Obsah",
      "text": "Vánoce už jsou za dveřmi! Proto se opět vydáme na vánoční výpravu do Nové Vsi pod Horou Svatého Šebestiána, která se bude konat **16. 12.–18. 12.**. Sejdeme se **na nádraží** v pátek 16. 12. **v 18:15** (vlak odjíždí v 18:40). Děti je třeba dopravit až na nádraží. Vrátíme se v neděli kolem **13:48** tamtéž.",
      "type": "normal",
    },
    Object {
      "color": "rgb(154, 21, 143)",
      "id": 3,
      "markdown": false,
      "name": "Patička",
      "text": "Janča, Domča a Lenča, vedoucí oddílu Sedmikrásky",
      "type": "footer",
    },
  ],
  "things": Array [
    Object {
      "id": 1,
      "name": "přezuvky",
      "selected": false,
    },
    Object {
      "id": 2,
      "name": "spacák",
      "selected": false,
    },
    Object {
      "id": 3,
      "name": "ešus",
      "selected": false,
    },
    Object {
      "id": 4,
      "name": "karimatku",
      "selected": false,
    },
    Object {
      "id": 5,
      "name": "věci na schůzku",
      "selected": false,
    },
    Object {
      "id": 6,
      "name": "zápisník a tužku",
      "selected": false,
    },
    Object {
      "id": 7,
      "name": "kroj",
      "selected": false,
    },
    Object {
      "id": 8,
      "name": "hygienu",
      "selected": false,
    },
    Object {
      "id": 9,
      "name": "baterku",
      "selected": false,
    },
    Object {
      "id": 10,
      "name": "večeři",
      "selected": false,
    },
    Object {
      "id": 11,
      "name": "šátek",
      "selected": false,
    },
  ],
}
`;
